src/main/java/com/palantir/gradle/versions/CheckBadPinsTask.java

```diff
@@ -0,0 +1,166 @@
+import com.google.common.collect.Sets;
+import com.palantir.gradle.versions.FuzzyPatternResolver.Glob;
+import com.palantir.gradle.versions.lockstate.Line;
+import com.palantir.gradle.versions.lockstate.LockState;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.nio.file.Files;
+import java.nio.file.StandardOpenOption;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.file.RegularFile;
+import org.gradle.api.file.RegularFileProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFile;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.options.Option;
+import org.gradle.language.base.plugins.LifecycleBasePlugin;
+
+public class CheckBadPinsTask extends DefaultTask {
+
+    private final Property<Boolean> shouldFix = getProject().getObjects().property(Boolean.class);
+    private final RegularFileProperty propsFileProperty =
+            getProject().getObjects().fileProperty();
+    private final RegularFileProperty lockFileProperty =
+            getProject().getObjects().fileProperty();
+
+    public CheckBadPinsTask() {
+        shouldFix.set(false);
+        setGroup(LifecycleBasePlugin.VERIFICATION_GROUP);
+        setDescription(
+                "Ensures matched versions in your versions.lock are pinned to avoid wasted dependency resolution.");
+        getOutputs().upToDateWhen(_task -> true); // task has no outputs, this is needed for it to be up to date
+    }
+
+    final void setLockFile(File lockFile) {
+        this.lockFileProperty.set(lockFile);
+    }
+
+    final void setPropsFile(File propsFile) {
+        this.propsFileProperty.set(propsFile);
+    }
+
+    @InputFile
+    public final Property<RegularFile> getLockFile() {
+        return lockFileProperty;
+    }
+
+    @InputFile
+    public final Property<RegularFile> getPropsFile() {
+        return propsFileProperty;
+    }
+
+    @Input
+    public final Property<Boolean> getShouldFix() {
+        return shouldFix;
+    }
+
+    @Option(option = "fix", description = "Whether to apply the suggested fix to versions.props")
+    public final void setShouldFix(boolean shouldFix) {
+        this.shouldFix.set(shouldFix);
+    }
+
+    @TaskAction
+    public final void checkBadPins() {
+        VersionsProps versionsProps =
+                VersionsProps.loadFromFile(getPropsFile().get().getAsFile().toPath());
+        LockState lockState =
+                new ConflictSafeLockFile(getLockFile().get().getAsFile().toPath()).readLocks();
+
+        checkBadPins(versionsProps, lockState);
+    }
+
+    private void checkBadPins(VersionsProps versionsProps, LockState lockState) {
```

Comment:
Might be possible to unit test just this method, so we can easily list all the possible examples.

Reply:
I'm not quite sure of the value of this over just using the integration spec - I have pulled out a static method `determineNewLines` which we can now test independently but I can only think of two relevant tests, versions props has all pins or versions props is missing pins - maybe there is something I've missed here but feel my tests just duplicate what is already being done

Reply:
It's only worth it if there are a lot of tests and we don't want to run the version that runs an entire Gradle every time.

Reply:
I'll leave both as I can see this getting enhancements in the future where more tests are required